internal/controller/apiextensions/composite/ready.go

```diff
@@ -67,6 +69,72 @@ type ReadinessCheck struct {
 
 	// MatchInt is the value you'd like to match if you're using "MatchInt" type.
 	MatchInteger *int64
+
+	// MatchCondition is the condition you'd like to match if you're using "MatchCondition" type.
+	// +optional
+	MatchCondition *MatchConditionReadinessCheck `json:"matchCondition,omitempty"`
+}
+
+// MatchConditionReadinessCheck is used to indicate how to tell whether a resource is ready
+// for consumption
+type MatchConditionReadinessCheck struct {
+	// Type indicates the type of condition you'd like to use.
+	// +kubebuilder:default="Ready"
+	Type string `json:"type,omitempty"`
+
+	// Status is the status of the condition you'd like to match.
+	// +kubebuilder:default="True"
+	Status string `json:"status,omitempty"`
+}
+
+// readinessCheckFromV1 derives a ReadinessCheck from the supplied v1.ReadinessCheck.
+func readinessCheckFromV1(in *v1.ReadinessCheck) ReadinessCheck {
+	if in == nil {
+		return ReadinessCheck{}
+	}
+
+	out := ReadinessCheck{
+		Type: ReadinessCheckType(in.Type),
+	}
+	if in.FieldPath != "" {
+		out.FieldPath = pointer.String(in.FieldPath)
+	}
+
+	// NOTE(negz): ComposedTemplate doesn't use pointer values for optional
+	// strings, so today the empty string and 0 are equivalent to "unset".
+	if in.MatchString != "" {
+		out.MatchString = pointer.String(in.MatchString)
+	}
+	if in.MatchInteger != 0 {
+		out.MatchInteger = pointer.Int64(in.MatchInteger)
+	}
+	if in.MatchCondition != nil {
+		out.MatchCondition = &MatchConditionReadinessCheck{
+			Type:   in.MatchCondition.Type,
+			Status: in.MatchCondition.Status,
+		}
+	}
+	return out
+}
+
+// readinessCheckFromDesired derives a ReadinessCheck from the supplied iov1alpha1.DesiredReadinessCheck.
+func readinessCheckFromDesired(in *iov1alpha1.DesiredReadinessCheck) ReadinessCheck {
```

Comment:
not sure whether to keep these private, everything else is public but used only in the package, so these might as well be public

Reply:
The public functions are a bad(?) habit of mine. 😬

I like to write everything as if it could be used as a public-facing library. I like that doing this:

* Encourages me to think about dependency injection, testability, etc.
* Requires me to write documentation for every function and method. (Linters complain if they're missing)

In practice we have everything under an `internal/` package, so only other packages inside of Crossplane _could_ use any of these public things.

I have a slight preference for keeping stuff public (inside) Crossplane, but I do admit it's a little bit of a "weird take". 😄 Perhaps here we could stick with the pattern of the rest of the file (i.e. public functions) and separately chat about whether we should be using more private functions and types?

Reply:
Ok, thanks for the explanation. Makes sense for me, lets keep them public.